src/meta/processors/job/JobManager.cpp

```diff
@@ -130,7 +130,12 @@ void JobManager::scheduleThread() {
     auto jobOp = std::get<0>(opJobId);
     auto jodId = std::get<1>(opJobId);
     auto spaceId = std::get<2>(opJobId);
-    std::lock_guard<std::recursive_mutex> lk(muJobFinished_[spaceId]);
+    auto iter = muJobFinished_.find(spaceId);
+    if (iter == muJobFinished_.end()) {
+      iter = muJobFinished_.emplace(spaceId, std::make_unique<std::recursive_mutex>()).first;
```

Comment:
Just curious, it seems that concurrent invoke on std::map.emplace() might also corrupt the internal RB tree?

Reply:
`std::mutex` cannot `move` and `copy`.There is a problem when calling the emplace interface of map

Reply:
Not very sure it's the same problem, it might not just be the problem of `std::mutex` `move()` && `copy()`, below is a piece of code that crashed and generated similar coredump as in issue https://github.com/vesoft-inc/nebula-ent/issues/1010:

```c++
#include <iostream>
#include <map>
#include <memory>
#include <mutex>
#include <string>
#include <thread>
#include <utility>

int main() {
  std::map<int, std::unique_ptr<std::recursive_mutex>> m;

  auto t1 = std::thread([&m] {
    for (int i = 0; i < 1000000; i++) {
      auto itr = m.find(i * 2);
      if (itr == m.end()) {
        m.emplace(
            std::make_pair(i * 2, std::make_unique<std::recursive_mutex>()));
      }
    }
  });

  auto t2 = std::thread([&m] {
    for (int i = 0; i < 1000000; i++) {
      auto itr = m.find(i * 2 + 1);
      if (itr == m.end()) {
        m.emplace(i * 2 + 1, std::make_unique<std::recursive_mutex>());
      }
    }
  });

  t1.join();
  t2.join();

  return 0;
}
```

```txt
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `./map'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f4669dbe4fa in std::_Rb_tree_insert_and_rebalance(bool, std::_Rb_tree_node_base*, std::_Rb_tree_node_base*, std::_Rb_tree_node_base&) ()
   from /lib/x86_64-linux-gnu/libstdc++.so.6
[Current thread is 1 (Thread 0x7f466999a700 (LWP 1859380))]
(gdb) bt
#0  0x00007f4669dbe4fa in std::_Rb_tree_insert_and_rebalance(bool, std::_Rb_tree_node_base*, std::_Rb_tree_node_base*, std::_Rb_tree_node_base&) ()
   from /lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x0000560133939086 in std::_Rb_tree<int, std::pair<int const, std::unique_ptr<std::recursive_mutex, std::default_delete<std::recursive_mutex> > >, std::_Select1st<std::pair<int const, std::unique_ptr<std::recursive_mutex, std::default_delete<std::recursive_mutex> > > >, std::less<int>, std::allocator<std::pair<int const, std::unique_ptr<std::recursive_mutex, std::default_delete<std::recursive_mutex> > > > >::_M_insert_node (this=0x7fff9296b6a0, __x=0x0,
    __p=0x7f4655b126f0, __z=0x7f4655b12760) at /usr/include/c++/9/bits/stl_tree.h:2366
#2  0x0000560133938889 in std::_Rb_tree<int, std::pair<int const, std::unique_ptr<std::recursive_mutex, std::default_delete<std::recursive_mutex> > >, std::_Select1st<std::pair<int const, std::unique_ptr<std::recursive_mutex, std::default_delete<std::recursive_mutex> > > >, std::less<int>, std::allocator<std::pair<int const, std::unique_ptr<std::recursive_mutex, std::default_delete<std::recursive_mutex> > > > >::_M_emplace_unique<std::pair<int, std::unique_ptr<std::recursive_mutex, std::default_delete<std::recursive_mutex> > > > (this=0x7fff9296b6a0) at /usr/include/c++/9/bits/stl_tree.h:2420
#3  0x0000560133938488 in std::map<int, std::unique_ptr<std::recursive_mutex, std::default_delete<std::recursive_mutex> >, std::less<int>, std::allocator<std::pair<int const, std::unique_ptr<std::recursive_mutex, std::default_delete<std::recursive_mutex> > > > >::emplace<std::pair<int, std::unique_ptr<std::recursive_mutex, std::default_delete<std::recursive_mutex> > > > (this=0x7fff9296b6a0) at /usr/include/c++/9/bits/stl_map.h:575
#4  0x0000560133937407 in <lambda()>::operator()(void) const (__closure=0x5601340bfeb8) at map.cpp:17
#5  0x0000560133937f98 in std::__invoke_impl<void, main()::<lambda()> >(std::__invoke_other, <lambda()> &&) (__f=...) at /usr/include/c++/9/bits/invoke.h:60
#6  0x0000560133937f02 in std::__invoke<main()::<lambda()> >(<lambda()> &&) (__fn=...) at /usr/include/c++/9/bits/invoke.h:95
#7  0x0000560133937e5c in std::thread::_Invoker<std::tuple<main()::<lambda()> > >::_M_invoke<0>(std::_Index_tuple<0>) (this=0x5601340bfeb8)
    at /usr/include/c++/9/thread:244
#8  0x0000560133937e06 in std::thread::_Invoker<std::tuple<main()::<lambda()> > >::operator()(void) (this=0x5601340bfeb8) at /usr/include/c++/9/thread:251
#9  0x0000560133937dd0 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<main()::<lambda()> > > >::_M_run(void) (this=0x5601340bfeb0)
    at /usr/include/c++/9/thread:195
#10 0x00007f4669dd2de4 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#11 0x00007f4669ee6609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#12 0x00007f4669c0e163 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb)
```

Reply:
good, this is caused by multi-threaded operation of map. . . `ConcurrentHashMap` should be used